### PR TITLE
test: add FSDP2 distributed coverage for AsyncGRPOTrainer

### PR DESCRIPTION
**Pull request by behroozazarkhalili**
`AsyncGRPOTrainer` had no distributed test coverage for training with FSDP2. I added a two-process functional test that uses a stub rollout worker and checks training steps, finite loss, and parameter updates on sharded parameters. I ran it successfully on two H100 GPUs and confirmed that it skips when fewer than two accelerators are available.

---

> [!NOTE]
> **Low Risk**
> Test-only changes with no production code paths; adds subprocess orchestration and GPU requirements for one experimental smoke test.
> 
> **Overview**
> Adds **2-process FSDP2 functional smoke coverage** for experimental `AsyncGRPOTrainer`, which previously only had single-process training tests.
> 
> A new **`test_train_fsdp2`** (marked slow, gated by `@require_torch_multi_accelerator`) spawns a self-contained companion script via `accelerate launch` and a dedicated **`fsdp2_reshard.yaml`** (2 ranks, FSDP v2, `fsdp_reshard_after_forward: true`). The worker runs two training steps with an in-process stub rollout queue and a **no-op weight transfer** (no vLLM / NCCL), then emits a parseable **`ASYNC_GRPO_FSDP2_RESULT`** JSON line. Pytest asserts all configured steps ran, loss is finite, parameters changed, and the launch really used FSDP2 sharding (2 processes, DTensor params).
> 
> The launcher pins **repo root on `PYTHONPATH`**, passes a temp **`output_dir`**, and on timeout **kills the full `/proc` descendant tree** so detached elastic ranks cannot hang the test.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 4eb9ba867171c8abe1a7dcabf87e3ee856f1d92b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by albertvillanova — COMMENTED** [review]
Thanks @behroozazarkhalili — this is careful work and the companion-worker pattern is a clean match for `_openreward_echo_env.py`. 

The blocker is that the test won't actually run anywhere in CI: the experimental lane runs on a single-GPU `aws-g4dn-2xlarge`, so `@require_torch_multi_accelerator` skips it (the green "Tests (experimental)" check here never executed it), and the only multi-GPU lane is the slow-tests `multi_gpu` job, which collects `-m slow` only — and this test isn't marked slow. Could you add `@pytest.mark.slow` so it lands in that lane (and confirm the runner really exposes 2 GPUs)? Otherwise it's a permanent skip that will quietly bitrot. 

Separately, the #6077 narrative in the module docstring and the test comment is very long for what it ultimately concludes — trimming to a one-line note would keep the test readable. The duplicated stub is fine given the subprocess boundary.

**Comment by behroozazarkhalili** [comment]
Thanks @albertvillanova, both points addressed in `2fc1fe06`.

**1. `@pytest.mark.slow` added.** You are right that it was a guaranteed permanent skip: `make slow_tests` is `pytest -m "slow" tests/`, so without the marker the test is deselected from both slow lanes, and the only other lane that collects it (`tests-experimental`) runs single-GPU where `@require_torch_multi_accelerator` skips it. With the marker it is now collectable by the `run_all_tests_multi_gpu` slow lane.

**2. Docstring/comment trimmed.** The long #6077 narrative is now a one-line note in both the test comment and the worker module docstring.

**On "confirm the runner really exposes 2 GPUs", being precise about what I could and could not verify from outside your infra:**

- What I confirmed from `slow-tests.yml`: the `run_all_tests_multi_gpu` job sets `CUDA_VISIBLE_DEVICES: "0,1"` (vs `"0"` for the single-GPU job), so it is configured for 2 GPUs.
- What I could not confirm: `torch.cuda.device_count() == 2` at runtime on that runner. I looked at the latest slow run (`28670435029`) and the `multi_gpu` and `single_gpu` jobs returned an identical `35 passed / 71 skipped`. That identity is not evidence of one GPU: it happens because there is currently no test that is both `@pytest.mark.slow` and `@require_torch_multi_accelerator`, so nothing in that run could differentiate the two lanes. This PR's test becomes the first such probe.

The nice property is that it is self-verifying and cannot go falsely red: `@require_torch_multi_accelerator` degrades to a visible **skip**, not a failure, when fewer than 2 accelerators are present. So the next slow-tests report will show either `test_train_fsdp2 PASSED` (confirming the multi_gpu runner really has 2 GPUs) or `SKIPPED` (flagging that the lane is effectively single-GPU and needs a real 2-GPU box), with no risk of a spurious failure either way.

If you would prefer a louder signal than a silent skip, I can add an explicit `assert torch.cuda.device_count() >= 2` in the worker so a misprovisioned runner fails loudly instead of quietly skipping. The tradeoff is that it would hard-fail on a 1-GPU runner rather than skip, so I left it as the standard skip guard and am happy to switch on your call.

cc @qgallouedec

File: tests/experimental/_async_grpo_fsdp2_worker.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+                yield RolloutSample(
+                    prompt=row["prompt"],
+                    completion=completions[idx],
+                    input_ids=prompt_ids + completion_ids,
+                    completion_mask=[0] * len(prompt_ids) + [1] * len(completion_ids),
+                    old_log_probs=[0.0] * len(prompt_ids) + [-0.5] * len(completion_ids),
+                    advantage=float(advantages[idx]),
+                    model_version=self._model_version,
+                    metrics={"reward": float(rewards[idx]), "reward_std": float(rewards.std())},
+                )
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply tests/experimental/_async_grpo_fsdp2_worker.py:104]: Confirmed and fixed in 002095df. `RolloutSample` gained a required `group_id` field while this branch was behind main, and the stub predated it, so the first sample raised `TypeError` before training started. The stub now assigns one group id per prompt, shared by all of its completions.

One caveat I want to be explicit about rather than imply this is verified: I could not execute the test locally. `AsyncGRPOTrainer` requires the `kernels-community/flash-attn3` kernel (padding-free training derives `cu_seq_lens` from `position_ids` resets, which SDPA and eager cannot express), and that kernel does not build on my machine. Since this test is marked `slow` plus `require_torch_multi_accelerator`, PR CI skips it too, so nothing in this pipeline has actually run it. The change is statically correct against the current dataclass, but it has not been executed.

File: tests/experimental/_async_grpo_fsdp2_worker.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        vllm_server_timeout=5.0,
+        report_to="none",
+    )
+    trainer = AsyncGRPOTrainer(
+        model=MODEL_ID,
+        reward_funcs=dummy_reward_func,
+        args=args,
+        train_dataset=dataset,
+        rollout_worker=_StubRolloutWorker(tokenizer, dataset, num_generations=3),
+    )
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply tests/experimental/_async_grpo_fsdp2_worker.py:176]: Confirmed and fixed in 002095df. Without an explicit `weight_transfer`, the trainer builds the default `WeightTransferClient`, which needs a live vLLM server, so the smoke was not self-contained as its docstring claimed. It now passes a no-op implementation of the protocol, and `send_weights` still drains the iterator so the weight-gathering path is exercised.

Same caveat as the sibling thread: I could not run this locally. `AsyncGRPOTrainer` hard-requires the `kernels-community/flash-attn3` kernel, which will not build on my machine, and PR CI skips this test because it is marked `slow` plus `require_torch_multi_accelerator`. So the fix is statically correct but unexecuted, and it would be worth a maintainer with flash-attn available running it before this merges.

File: tests/experimental/_async_grpo_fsdp2_worker.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        output_dir="async_grpo_fsdp2_out",
+        learning_rate=0.1,
+        per_device_train_batch_size=3,
+        num_generations=3,
+        max_completion_length=8,
+        token_budget=256,  # set explicitly; the stub worker has no real vLLM server to query for max_model_len
+        max_steps=2,
+        vllm_server_timeout=5.0,
+        report_to="none",
+    )
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply tests/experimental/_async_grpo_fsdp2_worker.py:166]: Confirmed. Driving TRL's own `TokenBudgetBatcher` with this worker's inputs reproduces the hang:

    10 samples x 41 tok, budget 256, 2 ranks -> rows [205, 205], 0 micro-batches
    24 samples,          budget 0,   2 ranks -> 4 micro-batches, no empty row

`token_budget=256` came from the sibling single-process `test_train`, where a single row of 256 overflows at the 7th sample. Two ranks double the capacity, so that overflow never happens here and the consumer blocks on an empty queue.

Fixed in 6c431cd9:

- `token_budget=0`, the only value that both selects the count-based `FixedCountBatcher` (`> 0` is false) and skips `wait_for_server_ready()` (`is None` is false), so the run still needs no vLLM server.
- `samples_per_weight_sync=24`, which covers `max_steps=2` with no weight-sync refill: micro-batch is `per_device_train_batch_size` 3 x 2 ranks = 6, so 2 steps need 12 samples and the previous 10 was short.
- `timeout=900` on `subprocess.run`, so a future starvation fails readably instead of hanging the pytest process.

File: tests/experimental/test_async_grpo_trainer.py [hunk 4]
```diff
@@ -309,6 +318,49 @@ def reset(self, **kwargs): ...
...
+        assert result.returncode == 0, f"FSDP2 worker failed:\nSTDOUT:\n{result.stdout}\nSTDERR:\n{result.stderr}"
+
+        result_lines = [ln for ln in result.stdout.splitlines() if ln.startswith(_FSDP2_RESULT_PREFIX)]
+        assert len(result_lines) == 1, f"expected exactly one result line, got {result_lines}\n{result.stdout}"
+        measured = json.loads(result_lines[0][len(_FSDP2_RESULT_PREFIX) :].strip())
+
+        # Training actually ran under FSDP2, produced a finite loss, and updated the parameters.
+        assert measured["steps"] >= 1, f"no training steps ran: {measured}"
+        assert measured["train_loss_finite"], f"train loss not finite under FSDP2: {measured}"
+        assert measured["params_changed"], f"parameters did not change under FSDP2: {measured}"
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply tests/experimental/test_async_grpo_trainer.py:391]: Confirmed, and fixed in 2433b1e3.

Verified each premise before changing anything:

- `AsyncGRPOTrainer` loads every model with `attn_implementation="kernels-community/flash-attn3"` (`async_grpo_trainer.py:811`), unconditionally.
- `AutoConfig.from_pretrained("trl-internal-testing/tiny-Qwen2ForCausalLM-2.5")` gives `hidden_size=8` over `num_attention_heads=4`, so `head_size=2`, not a multiple of 8.
- The FSDP2 worker drives that same model (`_async_grpo_fsdp2_worker.py:44`).
- The three siblings in this file carry the #6837 xfail; `test_train_fsdp2` did not.

It now carries the same marker, word for word. `xfail_strict` is unset, so an XPASS on a runner where FA accepts head_size=2 reports rather than fails, which matches the siblings.

The same commit corrects the comment above the test. It claimed `@pytest.mark.slow` routes the test to the multi-GPU `slow_tests` lane. It does not: `slow_tests` is `pytest -m "slow" tests/`, and `norecursedirs` excludes `tests/experimental` from that recursive collection.

    pytest -m slow tests/     --collect-only -> 0 matches for test_train_fsdp2
    pytest tests/experimental --collect-only -> 1 match

**Comment by behroozazarkhalili** [comment]
@albertvillanova a correction to what I told you on 2026-07-03, since you adjusted your review on the strength of it. It changes the conclusion rather than a detail, so the short version first: `test_train_fsdp2` has never executed in any lane since it was added, and it cannot execute today for three independent reasons.

I wrote: "With the marker it is now collectable by the `run_all_tests_multi_gpu` slow lane." That is wrong on all three counts.

The slow lane never collects it. `run_all_tests_multi_gpu` runs `make slow_tests`, which is `pytest -m "slow" tests/`, a recursive collection from `tests/`, and `pyproject.toml` sets `norecursedirs = ["tests/experimental", "tests/invariant"]`, so everything under `tests/experimental` drops out before the marker is consulted:

    pytest -m slow tests/     --collect-only -> 0 matches for test_train_fsdp2
    pytest tests/experimental --collect-only -> 1 match

The slow workflow would also never fire for this change. `slow-tests.yml` triggers on `push` to `main` with paths filtered to `trl/**.py` and `examples/**.py`, and a file under `tests/` matches neither, so merging this PR alone would not start that lane even if collection succeeded.

The lane that does collect it cannot run it. `test_experimental` is `pytest -n auto -s -v tests/experimental`, an explicit path, so `norecursedirs` does not apply, and what that lane reports is the answer to the question I twice said I could not settle from outside your infra:

    [gw16] SKIPPED tests/experimental/test_async_grpo_trainer.py::TestAsyncGRPOTrainer::test_train_fsdp2

That is [run 32436517199](https://github.com/huggingface/trl/actions/runs/32436517199), from today, and the same line appears in all six completed `Tests (experimental)` runs on this branch, back to the first two on 2026-06-24. Since `require_torch_multi_accelerator` skips rather than fails, a SKIP means the runner exposes fewer than two accelerators. That is as far as I can take it, because the guard reports no count and I cannot see your runner.

#6837 would stop the test even on a two-GPU lane. `AsyncGRPOTrainer` loads every model with `attn_implementation="kernels-community/flash-attn3"`, and `trl-internal-testing/tiny-Qwen2ForCausalLM-2.5` has hidden_size 8 over 4 attention heads, so head_size is 2 and Flash Attention rejects the forward pass. I marked the test `xfail` in 2433b1e3 to match the three siblings you already have marked, and corrected the in-source comment that repeated my wrong lane claim.

Because that skip is silent, the same trap stays open for whatever is added next to it. `Tests (experimental)` reports `568 passed, 9 skipped, 8 xfailed` and shows green with this test among the 9, so any future test carrying both `@pytest.mark.slow` and `@require_torch_multi_accelerator` under `tests/experimental` reaches only a lane that cannot run it, skips, and leaves the check green.

Which way to fix that is a repo-layout call. I would rather hand you the measurement than a recommendation I have not tested. Adding `tests/experimental` to the `slow_tests` target would admit every other slow-marked experimental test at once, at a runtime and pass-state I have not checked, and moving this test to a directory the multi-GPU lane already collects would separate it from the async-GRPO suite it belongs to. I am glad to do either, or to drop the multi-GPU test from this PR and keep the single-process coverage, if you would rather settle the lane question on its own.

File: tests/experimental/_async_grpo_fsdp2_worker.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+
+def main() -> None:
+    dataset = load_dataset("trl-internal-testing/zen", "conversational_prompt_completion", split="train")
+    tokenizer = AutoTokenizer.from_pretrained(MODEL_ID)
+
+    # Same minimal, memory-frugal config as the existing single-process test_train, with 2 steps so we
+    # exercise the optimizer loop more than once under FSDP2.
+    args = AsyncGRPOConfig(
+        output_dir="async_grpo_fsdp2_out",
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply tests/experimental/_async_grpo_fsdp2_worker.py]: Fixed in c75c6596.

The worker now takes its `output_dir` from the environment and the test passes `self.tmp_dir`, matching the five sibling tests in this file. Nothing is written to the working tree any more.

No default for the variable: this worker is launched only by `test_train_fsdp2`, so a missing value should fail loudly rather than silently fall back to the old relative path.

File: tests/experimental/test_async_grpo_trainer.py [hunk 6]
```diff
@@ -303,6 +313,62 @@ def reset(self, **kwargs): ...
...
+            stderr=subprocess.PIPE,
+            text=True,
+            start_new_session=True,
+        )
+        try:
+            stdout, stderr = proc.communicate(timeout=900)
+        except subprocess.TimeoutExpired:
+            os.killpg(proc.pid, signal.SIGKILL)
+            stdout, stderr = proc.communicate()
+            pytest.fail(f"FSDP2 worker timed out after 900s:\nSTDOUT:\n{stdout}\nSTDERR:\n{stderr}")
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply tests/experimental/test_async_grpo_trainer.py:377]: Confirmed, and it was my fix from the previous round that was wrong: killing the launcher's process group cannot reach ranks that torch elastic started with `start_new_session=True`, so a hang would have left them on the GPUs with the stdout pipe open and the drain would never have returned. The descendants are now read from /proc while the launcher is still their parent, and the launcher plus every descendant get SIGKILL before the pipes are drained. Reproduced with a launcher whose child detached into its own session: the child is found, `communicate()` returns at once after the kill, and no process survives.
